src/libsyntax_pos/lib.rs

```diff
@@ -962,34 +965,72 @@ impl FileMap {
             external_src: Lock::new(ExternalSource::Unneeded),
             start_pos,
             end_pos: Pos::from_usize(end_pos),
-            lines: Lock::new(Vec::new()),
-            multibyte_chars: Lock::new(Vec::new()),
-            non_narrow_chars: Lock::new(Vec::new()),
+            lines,
+            multibyte_chars,
+            non_narrow_chars,
             name_hash,
         }
     }
 
-    /// EFFECT: register a start-of-line offset in the
-    /// table of line-beginnings.
-    /// UNCHECKED INVARIANT: these offsets must be added in the right
-    /// order and must be in the right places; there is shared knowledge
-    /// about what ends a line between this file and parse.rs
-    /// WARNING: pos param here is the offset relative to start of CodeMap,
-    /// and CodeMap will append a newline when adding a filemap without a newline at the end,
-    /// so the safe way to call this is with value calculated as
-    /// filemap.start_pos + newline_offset_relative_to_the_start_of_filemap.
-    pub fn next_line(&self, pos: BytePos) {
-        // the new charpos must be > the last one (or it's the first one).
-        let mut lines = self.lines.borrow_mut();
-        let line_len = lines.len();
-        assert!(line_len == 0 || ((*lines)[line_len - 1] < pos));
-        lines.push(pos);
+    fn find_newlines_and_special_chars(src: &str, filemap_start_pos: BytePos)
+        -> (Vec<BytePos>, Vec<MultiByteChar>, Vec<NonNarrowChar>) {
+
+        let mut index = 0;
+        let mut lines = vec![filemap_start_pos];
+        let mut multibyte_chars = vec![];
+        let mut non_narrow_chars = vec![];
+
+        while index < src.len() {
+            let byte_pos = BytePos::from_usize(index) + filemap_start_pos;
+            let byte = src.as_bytes()[index];
+
+            if byte.is_ascii() {
+                match byte {
+                    b'\n' => {
+                        lines.push(byte_pos);
```

Comment:
This checks if the current character is a newline and then records the current position. Is there where the off-by one on the failing UI tests is coming from?

Reply:
Yes, I think that's the problem. Good catch!